generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/VersionResolver.java

```diff
@@ -1,47 +1,117 @@
-      if (restriction.isUpperBoundInclusive()) {
-        return restriction.getUpperBound().toString();
-      } else if (restriction.isLowerBoundInclusive()) {
-        return restriction.getLowerBound().toString();
+  /**
+   * Scrapes maven page for an artifact for all available versions, and returns a list of versions
+   * that meet the requirements provided by the version specifications.
+   *
+   * One important property of maven artifact page is that all versions are sorted from earliest to latest.
+   */
+  // TODO(petros): add support for repositories other than Maven central.
+  private List<String> getPotentialVersions(String groupId, String artifactId, String versionSpec)
+      throws IOException, Resolver.InvalidArtifactCoordinateException {
+    List<String> hyperLinks = downloader.scrapeHyperLinksForArtifact(groupId, artifactId);
+
+    List<String> potentialVersions = new ArrayList<>();
+    for (String link : hyperLinks) {
+      String potentialVersion = removeForwardSlashAtEnd(link);
+      if (isSnapshotVersion(potentialVersion)
+          || !containedInSpecifiedRange(groupId, artifactId, versionSpec, potentialVersion)) {
+        continue;
       }
-      // All bounds were exclusive.
+      potentialVersions.add(potentialVersion);
+    }
+    if (potentialVersions.isEmpty()) {
+      throw new Resolver.InvalidArtifactCoordinateException(
+          messageForInvalidArtifact(groupId, artifactId, versionSpec));
     }
+    return potentialVersions;
+  }
+
+  /** Checks if a given version is within the bounds of a version range */
+  private static boolean containedInSpecifiedRange(
+      String groupId, String artifactId, String versionSpec, String potentialVersion)
+      throws Resolver.InvalidArtifactCoordinateException {
+
+    if (versionSpec.isEmpty()) {
+      return true;
+    }
+    versionSpec = transformBasicVersionIntoRange(versionSpec);
+    VersionRange range;
+    try {
+      range = VersionRange.createFromVersionSpec(versionSpec);
+    } catch (InvalidVersionSpecificationException e) {
+      throw new Resolver.InvalidArtifactCoordinateException(
+          messageForInvalidArtifact(groupId, artifactId, versionSpec));
+    }
+    return range.containsVersion(new DefaultArtifactVersion(potentialVersion));
+  }
+
+  /** Transforms a basic version spec, i.e. "1.3.4" into "[1.3.4, )" */
+  private static String transformBasicVersionIntoRange(String versionSpec) {
+    if (versionSpec.contains("[") || versionSpec.contains("(")) {
+      // Already a version range.
+      return versionSpec;
+    }
+    return "[" + versionSpec + "," + ")";
+  }
+
+  /** Checks if a version is a snapshot versions, i.e. non-release versions */
+  // TODO add support for sequences of versions like "[2.0,2.4),(2.5,)"
+  private static boolean isSnapshotVersion(String version) {
+    return !Pattern.compile("\\d+(\\.\\d+)*(\\.\\d+)*").matcher(version).matches() || version.isEmpty();
```

Comment:
I chose not to implement support for version ranges like "(2.0,2.4),(2.5,)." The regular expression would be very gnarly.  The only alternative I could think of was defining a small grammar and using ANTLR to generate a parser. That seemed rather excessive though. Thoughts?